src/node_http2.cc

```diff
@@ -908,300 +1154,920 @@ void Http2Session::OnTrailers(Nghttp2Stream* stream,
+}
+
+ssize_t Http2Stream::Provider::Stream::OnRead(nghttp2_session* handle,
+                                              int32_t id,
+                                              uint8_t* buf,
+                                              size_t length,
+                                              uint32_t* flags,
+                                              nghttp2_data_source* source,
+                                              void* user_data) {
+  Http2Session* session = static_cast<Http2Session*>(user_data);
+  DEBUG_HTTP2SESSION2(session, "reading outbound data for stream %d", id);
+  Http2Stream* stream = GetStream(session, id, source);
+  CHECK_EQ(id, stream->id());
+
+  size_t amount = 0;          // amount of data being sent in this data frame.
+
+  uv_buf_t current;
+
+  if (!stream->queue_.empty()) {
+    DEBUG_HTTP2SESSION2(session, "stream %d has pending outbound data", id);
+    nghttp2_stream_write* head = stream->queue_.front();
+    current = head->bufs[stream->queue_index_];
+    size_t clen = current.len - stream->queue_offset_;
+    amount = std::min(clen, length);
+    DEBUG_HTTP2SESSION2(session, "sending %d bytes for data frame on stream %d",
+                        amount, id);
+    if (amount > 0) {
+      memcpy(buf, current.base + stream->queue_offset_, amount);
+      stream->queue_offset_ += amount;
     }
-    // For performance, we pass name and value pairs to array.protototype.push
-    // in batches of size NODE_PUSH_VAL_TO_ARRAY_MAX * 2 until there are no
-    // more items to push.
-    if (j > 0) {
-      fn->Call(env()->context(), holder, j * 2, argv).ToLocalChecked();
+    if (stream->queue_offset_ == current.len) {
+      stream->queue_index_++;
+      stream->queue_offset_ = 0;
+    }
+    if (stream->queue_index_ == head->nbufs) {
+      head->cb(head->req, 0);
+      delete head;
+      stream->queue_.pop();
+      stream->queue_offset_ = 0;
+      stream->queue_index_ = 0;
     }
   }
 
-  Local<Value> args[4] = {
-    Integer::New(isolate, stream->id()),
-    Integer::New(isolate, cat),
-    Integer::New(isolate, flags),
-    holder
-  };
-  MakeCallback(env()->onheaders_string(), arraysize(args), args);
+  if (amount == 0 && stream->IsWritable() && stream->queue_.empty()) {
+    DEBUG_HTTP2SESSION2(session, "deferring stream %d", id);
+    return NGHTTP2_ERR_DEFERRED;
+  }
+
+  if (stream->queue_.empty() && !stream->IsWritable()) {
+    DEBUG_HTTP2SESSION2(session, "no more data for stream %d", id);
+    *flags |= NGHTTP2_DATA_FLAG_EOF;
+
+    session->GetTrailers(stream, flags);
+  }
+
+  return amount;
+}
+
+
+
+// Implementation of the JavaScript API
+
+void HttpErrorString(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+  uint32_t val = args[0]->Uint32Value(env->context()).ToChecked();
+  args.GetReturnValue().Set(
+      OneByteString(env->isolate(), nghttp2_strerror(val)));
```

Comment:
`OneByteString()` uses `v8::NewStringType::kNormal` – can you use `kInternalized`, since these are static strings (i.e. call `NewFromOneByte` directly with that flag)?

Reply:
Good call :)